lib/puppet/type/cisco_vni.rb

```diff
@@ -0,0 +1,85 @@
+# Manages a Cisco VNI.
+#
+# December 2015
+#
+# Copyright (c) 2013-2015 Cisco and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+Puppet::Type.newtype(:cisco_vni) do
+  @doc = "Manages a Cisco VNI.
+
+  cisco_vni {\"<vni>\":
+    ..attributes..
+  }
+
+  <vni> is the id of the vni.
+
+  Example:
+    cisco_vni {\"10000\":
+      ensure      => present,
+      mapped_vlan => 100,
+    }"
+
+  ###################
+  # Resource Naming #
+  ###################
+
+  # Parse out the title to fill in the attributes in these
+  # patterns. These attributes can be overwritten later.
+  def self.title_patterns
+    identity = ->(x) { x }
+    patterns = []
+
+    # Below pattern matches both parts of the full composite name.
+    patterns << [
+      /^(\d+)$/,
+      [
+        [:vni, identity]
+      ],
+    ]
+    patterns
+  end
+
+  newparam(:vni, namevar: true) do
+    desc 'ID of the Virtual Network Identifier. Valid values are integer.'
+
+    validate do |id|
+      valid_ids = *(4096..1_677_721_5)
+
+      unless valid_ids.include?(id.to_i)
+        fail('ID is not in the valid range.')
+      end # if
+    end
+  end # param id
+
+  ##############
+  # Attributes #
+  ##############
+
+  ensurable
+
+  newproperty(:mapped_vlan) do
+    desc 'ID of the VLAN mapped to teh VNI. Valid values are integer.'
+    # Keyword 'default' is not supported as there is no default mapping
```

Comment:
Well I had this earlier and because of this ran into a problem of idempotence. 
cisco_vni {'10000':
ensure => 'present'
 mapped_vlan => 'default'
}
In the earlier implementation this would destroy the vlan to vni mapping. But when you apply this manifest again and again, every time this would create a resource which idempotence beaker test doesn't like. This is because actually there is no 'default' mapping. There is either a defined mapping or none, but no default.
Chris and I had a discussion about it and decided to get rid of 'default' keyword. Your thoughts?